compiler/src/iree/compiler/InputConversion/TOSA/TosaToLinalgExt.cpp

```diff
@@ -0,0 +1,178 @@
+    });
+
+    indices = builder.create<tensor::ExpandShapeOp>(
+        indicesTy.clone(expandIndShape), indices, expandIndMap);
+    indicesTy = indices.getType().dyn_cast<RankedTensorType>();
+
+    // Materialize the batch indice as LinalgExt scatter is not batched.
+    {
+      llvm::SmallVector<Value> dynDims;
+      for (int i = 0, s = indicesTy.getRank(); i < s; ++i)
+        if (indicesTy.isDynamicDim(i))
+          dynDims.push_back(builder.create<tensor::DimOp>(indices, i));
+
+      Value empty = builder.create<tensor::EmptyOp>(
+          indicesTy.getShape(), indicesTy.getElementType(), dynDims);
+
+      Value batchIdx = nullptr;
+
+      if (indicesTy.getDimSize(0) == 1) {
+        Value zero = builder.create<arith::ConstantOp>(
+            rewriter.getZeroAttr(indicesTy.getElementType()));
+        batchIdx = builder.create<linalg::FillOp>(zero, empty).getResult(0);
+      } else {
+        SmallVector<utils::IteratorType> iterators(
+            indicesTy.getRank(), utils::IteratorType::parallel);
+        SmallVector<AffineMap, 3> indexingMaps(
+            2, builder.getMultiDimIdentityMap(indicesTy.getRank()));
+        batchIdx = builder
+                       .create<linalg::GenericOp>(
+                           indicesTy, indices, empty, indexingMaps, iterators,
+                           [&](OpBuilder &nestedBuilder, Location nestedLoc,
+                               ValueRange blockArgs) {
+                             ImplicitLocOpBuilder b(op.getLoc(), nestedBuilder);
+                             auto index = b.create<linalg::IndexOp>(0);
+                             auto cast = b.create<arith::IndexCastOp>(
+                                 indicesTy.getElementType(), index);
+                             b.create<linalg::YieldOp>(cast.getResult());
+                           })
+                       .getResult(0);
+      }
+
+      indicesTy =
+          indicesTy.clone({indicesTy.getDimSize(0), indicesTy.getDimSize(1), 2})
+              .cast<RankedTensorType>();
+      indices = builder.create<tosa::ConcatOp>(indicesTy,
+                                               ValueRange{batchIdx, indices},
+                                               rewriter.getI64IntegerAttr(2));
+    }
+
+    auto collapseBatch = [](Value value, ImplicitLocOpBuilder &b) -> Value {
+      auto valueTy = value.getType().cast<ShapedType>();
+      llvm::SmallVector<int64_t> collapseShape(valueTy.getShape().drop_front());
+      llvm::SmallVector<ReassociationExprs, 4> collapseMap(valueTy.getRank() -
+                                                           1);
+      collapseMap.front().push_back(b.getAffineDimExpr(0));
+      for (int i = 0, s = collapseMap.size(); i < s; ++i) {
+        collapseMap[i].push_back(b.getAffineDimExpr(i + 1));
+      }
+
+      int64_t batch = valueTy.getShape().front();
+      int64_t rows = collapseShape.front();
+      int64_t batchDyn = batch == ShapedType::kDynamic;
```

Comment:
Assigning a bool to an int64_t feils weird, seems like you are using it as a bool below)

Reply:
Done.